### ISSUE TITLE
Ability to push to packit's dist-git fork

### ISSUE THREAD

**Issue by nforro** [packit/packit-service#1926]
As a Fedora maintainer I have pull-from-upstream configured for a package.

After a new upstream release packit sucessfully creates PRs in dist-git, for instance:
https://src.fedoraproject.org/rpms/gpxsee/pull-request/4

However, the build fails because the package gained a new dependency. It would be a trivial fix, just adding a new `Requires` line to the spec file, but I'm not able to push to the PR branch. Which means I can't take advantage of the PR and packit's downstream automation, I have to do everything manually.

Please correct me if I'm wrong, maybe I'm missing something.

**Comment by lachmanfrantisek** [comment]
Yes, that's correct... and being able to push to a fork would be really nice. (E.g. via an argument)

Another option would be to be able to rerun `propose-downstream` from an upstream git ref (E.g. via a comment argument). So, you would fix this upstream first and rerun the job. Currently, you need to do a new release after the fix.


---

Locally, you can fetch the source branch of the dist-git PR and push it (with a fix) to your fork. You can maybe redefine the dist-git URL in a Packit config and regenerate the PRs.

For local usage, we can either add a CLI option (e.g. `--push-to-fork`) to push to the fork, or a CLI option (e.g. `--dist-git-url`) to redefine the dist-git project on the go (The latter would be also handy to push to a different dist-git instance...)

**Comment by nforro** [comment]
> Another option would be to be able to rerun `propose-downstream` from an upstream git ref (E.g. via a comment argument). So, you would fix this upstream first and rerun the job. Currently, you need to do a new release after the fix.

But in case of pull-from-upstream there would be no such workaround.

> Locally, you can fetch the source branch of the dist-git PR and push it (with a fix) to your fork. You can probably redefine the dist-git URL in a Packit config and regenerate the PRs.

That's true, but not merging the original PR means `koji_build` and `bodhi_update` jobs won't be triggered, or am I wrong?

**Comment by lbarcziova** [comment]
I agree that being able to push to the branch with the changes would be the best (both for `propose-downsteam` and `pull-from-upstream`), I believe with `propose-downstream` we already mentioned this when we had to do some changes on top of the automatic PRs. I am not sure whether this can be configured in Pagure somehow :thinking:

But as Franta wrote, currently the easiest thing for users to do here is to locally checkout the branch created by Packit, fix it and push it to your fork. Should we for the beginning at least document this?

>  That's true, but not merging the original PR means koji_build and bodhi_update jobs won't be triggered, or am I wrong?

This behaviour can be changed by configuring  `allowed_pr_authors` and `allowed_committers` (https://packit.dev/docs/configuration/#koji_build) for `koji_build` job.

**Comment by lachmanfrantisek** [comment]
Laura is quicker with the configs... but you can also use comment commands in the PR to trigger that manually...

**Comment by lachmanfrantisek** [comment]
I've checked the Pagure settings both on src.fedoraproject.org and pagure.io and haven't found anything... And nothing in the pagure's issue tracker... We can maybe ask there.

**Comment by nforro** [comment]
> This behaviour can be changed by configuring `allowed_pr_authors` and `allowed_committers` (https://packit.dev/docs/configuration/#koji_build) for `koji_build` job.

Thanks, yes, that's a good point. However, that also enables it for every direct push to dist-git. That's not a bad thing, but I can imagine it's not what a maintainer would always want.

> But as Franta wrote, currently the easiest thing for users to do here is to locally checkout the branch created by Packit, fix it and push it to your fork. Should we for the beginning at least document this?

It would be nice to have this mentioned somewhere (perhaps as a tip in pull-from-upstream documentation?).

**Comment by lbarcziova** [comment]
>  Thanks, yes, that's a good point. However, that also enables it for every direct push to dist-git. That's not a bad thing, but > I can imagine it's not what a maintainer would always want.

That's not completely right; we react only to pushes with a specfile change.

> It would be nice to have this mentioned somewhere (perhaps as a tip in pull-from-upstream documentation?).

+1, I will add it to my existing PR

**Comment by nforro** [comment]
> That's not completely right; we react only to pushes with a specfile change.

Hm, thinking about that, what if there is a change in the spec but no release bump? That happens sometimes, for minor changes that don't require an immediate rebuild. I think Koji refuses to build the same NVR twice.

**Comment by nforro** [comment]
Yes, that would be nice. Actually, looking at the commands, they could be simplified a little bit (fetching via https rather than ssh), eliminating one variable:

```bash
git fetch https://src.fedoraproject.org/forks/packit/rpms/$YOUR_PACKAGE.git refs/heads/*:refs/remotes/packit/*
git cherry-pick packit/$VERSION-$BRANCH-update-pull_from_upstream
```

**Comment by lbarcziova** [comment]
I created a separate issue: packit/packit#1915.

Can we close this now, or do we want to have this as a tracker of the Pagure issue?
